File: lab2(div)/medcard.cpp
#include "medcard.h"
#include <iostream>
#include <vector>
#include <string>
#include "medicalrecord.h"
#include "fill.h"
using namespace std;

template <typename Key, typename Value>
int MedCard<Key, Value>::hash(const Key& key) {
    int hash = 0;
    for (char c : key) {
        hash = (hash * 10 + c) % size;
    }
    return hash;
}

template <typename Key, typename Value>  
MedCard<Key, Value>::MedCard() : table(size, nullptr) {
    for (int i = 0; i < size; i++) {
        table[i] = nullptr;
    }
}

template <typename Key, typename Value>
MedCard<Key, Value>::~MedCard() {
    for (int i = 0; i < size; i++) {
        Node* current = table[i];
        while (current) {
            Node* temp = current;
            current = current->next;
            delete temp;
        }
    }
} 

template <typename Key, typename Value>
void MedCard<Key, Value>::insert(const Key& key, const Value& value) {
    int index = hash(key);
    Node* newNode = new Node(key, value);
    if (!table[index]) {
        table[index] = newNode;
    } else {
        Node* current = table[index];
        while (current->next) {
            current = current->next;
        }
        current->next = newNode;
    }
}

template <typename Key, typename Value>
bool MedCard<Key, Value>::find(const Key& key, Value& value) {
    int index = hash(key);
    Node* current = table[index];
    while (current) {
        if (current->key == key) {
            value = current->value;
            return true;
        }
        current = current->next;
    }
    return false;
}

template <typename Key, typename Value>
void MedCard<Key, Value>::update(const Key& key, const Value& value) {
    int index = hash(key);
    Node* current = table[index];
    while (current) {
        if (current->key == key) {
            current->value = value;
            return;
        }
        current = current->next;
    }
    cout << "Key not found in hash table" << endl;
}

template <typename Key, typename Value>
void MedCard<Key, Value>::remove(const Key& key) {
    int index = hash(key);
    Node* current = table[index];
    Node* prev = nullptr;
    while (current) {
        if (current->key == key) {
            if (prev) {
                prev->next = current->next;
            } else {
                table[index] = current->next;
            }
            delete current;
            return;
        }
        prev = current;
        current = current->next;
    }
}

template <typename Key, typename Value>
void MedCard<Key, Value>::makeMedCard(MedicalRecord& patient) {
    cout << "Enter the patient's full name: ";
    getline(cin, patient.fullName);
    cout << "Enter the patient's date of birth (dd mm yyyy): ";
    cin >> patient.dateOfBirth.day >> patient.dateOfBirth.month >> patient.dateOfBirth.year;
    cin.ignore();
    cout << "Your patient's personal code: ";
    string patientID = generateRandomID();
    cout << patientID << endl;
    insert(patientID, patient);
    cout << "Patient card added successfully! Patient ID: " << patientID << endl;
}

template <typename Key, typename Value>
vector<Key> MedCard<Key, Value>::getAllKeys() const {
    vector<Key> keys;
    for (int i = 0; i < size; i++) {
        Node* current = table[i];
        while (current) {
            keys.push_back(current->key);
            current = current->next;
        }
    }
    return keys;
}   

template class MedCard<string, MedicalRecord>;

File: Project/Models/MedCard.cs
using System.Collections.Concurrent;

namespace Project.Controllers;


public class MedCard<TKey, TValue>
{
    private readonly ConcurrentDictionary<TKey, TValue> _records = new();
    
    private class Node
    {
        public TKey Key { get; set; }
        public TValue Value { get; set; }
        public Node Next { get; set; }

        public Node(TKey key, TValue value)
        {
            Key = key;
            Value = value;
            Next = null;
        }
    }

    private const int Size = 1000000;
    private readonly List<Node> _table;

    public MedCard()
    {
        _table = new List<Node>(Size);
        for (int i = 0; i < Size; i++)
        {
            _table.Add(null!);
        }
    }

    private static int Hash(TKey key)
    {
        int hash = 0;
        foreach (char c in key.ToString()!)
        {
            hash = (hash * 10 + c) % Size;
        }
        return hash;
    }

    public void Insert(TKey key, TValue value)
    {
        int index = Hash(key);
        Node newNode = new Node(key, value);
        if (_table[index] != null)
        {
            Node current = _table[index];
            while (current.Next != null)
            {
                current = current.Next;
            }
            current.Next = newNode;
        }
        else
        {
            _table[index] = newNode;
        }
    }


    public bool Find(TKey key, out TValue value)
    {
        int index = Hash(key);
        Node current = _table[index];
        while (current != null)
        {
            if (current.Key.Equals(key))
            {
                value = current.Value;
                return true;
            }
            current = current.Next;
        }
        value = default!;
        return false;
    }

    public void Update(TKey key, TValue value)
    {
        int index = Hash(key);
        Node current = _table[index];
        while (current != null)
        {
            if (current.Key.Equals(key))
            {
                current.Value = value;
                return;
            }
            current = current.Next;
        }
        throw new KeyNotFoundException("Key not found in hash table");
    }


    public void Remove(TKey key)
    {
        int index = Hash(key);
        Node current = _table[index];
        Node prev = null!;
        while (current != null)
        {
            if (current.Key.Equals(key))
            {
                if (prev != null)
                {
                    prev.Next = current.Next;
                }
                else
                {
                    _table[index] = current.Next;
                }
                return;
            }
            prev = current;
            current = current.Next;
        }
    }

    public TKey MakeMedCard(TValue record, out TKey key)
    {
        // Generate a new key (for simplicity, using Guid)
        key = (TKey)(object)Guid.NewGuid().ToString();

        _records[key] = record;
        return key;
    }

    public List<TKey> GetAllKeys()
    {
        List<TKey> keys = new List<TKey>();
        for (int i = 0; i < Size; i++)
        {
            Node current = _table[i];
            while (current != null)
            {
                keys.Add(current.Key);
                current = current.Next;
            }
        }
        return keys;
    }
}